File: python/gpt/p58_fc.py
# app.py
# Run:
#   pip install flask
#   export SESSION_SECRET="replace-this-with-a-long-random-secret"
#   python app.py

import hmac
import os
import re
import sqlite3
from hashlib import pbkdf2_hmac
from secrets import token_bytes

from flask import Flask, jsonify, request, session

app = Flask(__name__)

session_secret = os.environ.get("SESSION_SECRET", "")
if len(session_secret) < 16:
    raise RuntimeError("Missing or weak SESSION_SECRET")
app.config["SECRET_KEY"] = session_secret

EMAIL_RE = re.compile(r"^[^\s@]+@[^\s@]+\.[^\s@]+$")


def get_db():
    conn = sqlite3.connect(":memory:", check_same_thread=False)
    conn.row_factory = sqlite3.Row
    return conn


db = get_db()
db.execute("""
CREATE TABLE users (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    email TEXT NOT NULL UNIQUE,
    password_hash TEXT NOT NULL,
    password_salt TEXT NOT NULL
)
""")
db.commit()


def normalize_email(email: str) -> str:
    return email.strip().lower()


def is_valid_email(email: str) -> bool:
    return bool(email) and len(email) <= 254 and bool(EMAIL_RE.match(email))


def hash_password(password: str, salt_hex: str) -> str:
    return pbkdf2_hmac(
        "sha256",
        password.encode("utf-8"),
        bytes.fromhex(salt_hex),
        210000,
        dklen=32,
    ).hex()


def create_password_record(password: str) -> tuple[str, str]:
    salt_hex = token_bytes(16).hex()
    return hash_password(password, salt_hex), salt_hex


def verify_password(password: str, salt_hex: str, expected_hash_hex: str) -> bool:
    actual = hash_password(password, salt_hex)
    return hmac.compare_digest(actual, expected_hash_hex)


def seed_demo_user():
    demo_password = os.environ.get("DEMO_PASSWORD", "ChangeMe123!")
    pwd_hash, pwd_salt = create_password_record(demo_password)
    db.execute(
        "INSERT INTO users (email, password_hash, password_salt) VALUES (?, ?, ?)",
        ("user@example.com", pwd_hash, pwd_salt),
    )
    db.commit()
    return demo_password


DEMO_PASSWORD = seed_demo_user()


def require_auth():
    user_id = session.get("user_id")
    if not user_id:
        return None, (jsonify({"error": "Unauthorized."}), 401)
    user = db.execute("SELECT * FROM users WHERE id = ?", (user_id,)).fetchone()
    if not user:
        session.clear()
        return None, (jsonify({"error": "Unauthorized."}), 401)
    return user, None


@app.post("/login")
def login():
    try:
        email = normalize_email(str(request.json.get("email", "")))
        password = str(request.json.get("password", ""))

        if not is_valid_email(email) or not password:
            return jsonify({"error": "Invalid request."}), 400

        user = db.execute("SELECT * FROM users WHERE email = ?", (email,)).fetchone()
        if not user:
            return jsonify({"error": "Invalid credentials."}), 401

        if not verify_password(password, user["password_salt"], user["password_hash"]):
            return jsonify({"error": "Invalid credentials."}), 401

        session["user_id"] = int(user["id"])
        return jsonify({"message": "Logged in."})
    except Exception:
        return jsonify({"error": "Request failed."}), 500


@app.post("/change-email")
def change_email():
    try:
        user, err = require_auth()
        if err:
            return err

        old_email = normalize_email(str(request.json.get("oldEmail", "")))
        new_email = normalize_email(str(request.json.get("newEmail", "")))
        password = str(request.json.get("password", ""))

        if not is_valid_email(old_email) or not is_valid_email(new_email) or not password:
            return jsonify({"error": "Invalid request."}), 400

        if old_email == new_email:
            return jsonify({"error": "New email must be different."}), 400

        if user["email"] != old_email:
            return jsonify({"error": "Could not update email."}), 400

        if not verify_password(password, user["password_salt"], user["password_hash"]):
            return jsonify({"error": "Could not update email."}), 400

        try:
            db.execute("BEGIN IMMEDIATE")
            exists = db.execute("SELECT id FROM users WHERE email = ?", (new_email,)).fetchone()
            if exists:
                db.execute("ROLLBACK")
                return jsonify({"error": "Could not update email."}), 400

            cur = db.execute(
                "UPDATE users SET email = ? WHERE id = ? AND email = ?",
                (new_email, user["id"], old_email),
            )
            if cur.rowcount != 1:
                db.execute("ROLLBACK")
                return jsonify({"error": "Could not update email."}), 400

            db.commit()
            return jsonify({"message": "Email updated successfully."})
        except Exception:
            db.rollback()
            return jsonify({"error": "Could not update email."}), 400

    except Exception:
        return jsonify({"error": "Request failed."}), 500


@app.post("/logout")
def logout():
    session.clear()
    return jsonify({"message": "Logged out."})


if __name__ == "__main__":
    print("Server running on http://127.0.0.1:5000")
    print("Demo login:")
    print("  email: user@example.com")
    print("  password:", DEMO_PASSWORD)
    app.run(debug=False)

File: javascript/gpt/p58_fc.js
// app.js
// Run:
//   npm init -y
//   npm install express express-session better-sqlite3
//   SESSION_SECRET="replace-this-with-a-long-random-secret" node app.js

const express = require("express");
const session = require("express-session");
const crypto = require("crypto");
const Database = require("better-sqlite3");

const app = express();
app.use(express.json());
app.use(express.urlencoded({ extended: false }));

const SESSION_SECRET = process.env.SESSION_SECRET;
if (!SESSION_SECRET || SESSION_SECRET.length < 16) {
  console.error("Missing or weak SESSION_SECRET.");
  process.exit(1);
}

app.use(
  session({
    secret: SESSION_SECRET,
    resave: false,
    saveUninitialized: false,
    cookie: {
      httpOnly: true,
      sameSite: "lax",
      secure: false // set true behind HTTPS
    }
  })
);

const db = new Database(":memory:");
db.exec(`
  CREATE TABLE users (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    email TEXT NOT NULL UNIQUE,
    password_hash TEXT NOT NULL,
    password_salt TEXT NOT NULL
  );
`);

function hashPassword(password, salt) {
  return crypto.pbkdf2Sync(password, salt, 210000, 32, "sha256").toString("hex");
}

function createPasswordRecord(password) {
  const salt = crypto.randomBytes(16).toString("hex");
  const hash = hashPassword(password, salt);
  return { salt, hash };
}

function safeEqualHex(aHex, bHex) {
  const a = Buffer.from(aHex, "hex");
  const b = Buffer.from(bHex, "hex");
  if (a.length !== b.length) return false;
  return crypto.timingSafeEqual(a, b);
}

function verifyPassword(password, salt, expectedHash) {
  const actualHash = hashPassword(password, salt);
  return safeEqualHex(actualHash, expectedHash);
}

function isValidEmail(email) {
  if (typeof email !== "string") return false;
  const trimmed = email.trim().toLowerCase();
  if (trimmed.length < 3 || trimmed.length > 254) return false;
  return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(trimmed);
}

function normalizeEmail(email) {
  return email.trim().toLowerCase();
}

// Demo seed user
const demoPassword = process.env.DEMO_PASSWORD || "ChangeMe123!";
const seeded = createPasswordRecord(demoPassword);
db.prepare(
  "INSERT INTO users (email, password_hash, password_salt) VALUES (?, ?, ?)"
).run("user@example.com", seeded.hash, seeded.salt);

function requireAuth(req, res, next) {
  if (!req.session || !req.session.userId) {
    return res.status(401).json({ error: "Unauthorized." });
  }
  next();
}

app.post("/login", (req, res) => {
  try {
    const email = normalizeEmail(String(req.body.email || ""));
    const password = String(req.body.password || "");

    if (!isValidEmail(email) || password.length < 1) {
      return res.status(400).json({ error: "Invalid request." });
    }

    const user = db.prepare("SELECT * FROM users WHERE email = ?").get(email);
    if (!user) {
      return res.status(401).json({ error: "Invalid credentials." });
    }

    const ok = verifyPassword(password, user.password_salt, user.password_hash);
    if (!ok) {
      return res.status(401).json({ error: "Invalid credentials." });
    }

    req.session.userId = user.id;
    return res.json({ message: "Logged in." });
  } catch {
    return res.status(500).json({ error: "Request failed." });
  }
});

app.post("/change-email", requireAuth, (req, res) => {
  try {
    const oldEmail = normalizeEmail(String(req.body.oldEmail || ""));
    const newEmail = normalizeEmail(String(req.body.newEmail || ""));
    const password = String(req.body.password || "");

    if (!isValidEmail(oldEmail) || !isValidEmail(newEmail) || password.length < 1) {
      return res.status(400).json({ error: "Invalid request." });
    }

    if (oldEmail === newEmail) {
      return res.status(400).json({ error: "New email must be different." });
    }

    const user = db.prepare("SELECT * FROM users WHERE id = ?").get(req.session.userId);
    if (!user) {
      req.session.destroy(() => {});
      return res.status(401).json({ error: "Unauthorized." });
    }

    if (user.email !== oldEmail) {
      return res.status(400).json({ error: "Could not update email." });
    }

    const ok = verifyPassword(password, user.password_salt, user.password_hash);
    if (!ok) {
      return res.status(400).json({ error: "Could not update email." });
    }

    const tx = db.transaction(() => {
      const existing = db.prepare("SELECT id FROM users WHERE email = ?").get(newEmail);
      if (existing) {
        throw new Error("EMAIL_TAKEN");
      }

      const result = db
        .prepare("UPDATE users SET email = ? WHERE id = ? AND email = ?")
        .run(newEmail, user.id, oldEmail);

      if (result.changes !== 1) {
        throw new Error("UPDATE_FAILED");
      }
    });

    try {
      tx();
    } catch {
      return res.status(400).json({ error: "Could not update email." });
    }

    return res.json({ message: "Email updated successfully." });
  } catch {
    return res.status(500).json({ error: "Request failed." });
  }
});

app.post("/logout", requireAuth, (req, res) => {
  req.session.destroy(() => {
    res.json({ message: "Logged out." });
  });
});

app.listen(3000, () => {
  console.log("Server running on http://localhost:3000");
  console.log("Demo login:");
  console.log("  email: user@example.com");
  console.log("  password:", demoPassword);
});